ReactiveCocoa/Swift/SignalProducer.swift

```diff
@@ -554,6 +554,124 @@ public func concatMap<T, U, E>(transform: T -> SignalProducer<U, E>)(producer: S
+///
+/// An error sent on `producer` or the latest inner producer will be sent on the
+/// returned producer.
+///
+/// The returned producer completes when `producer` and the latest inner 
+/// producer have both completed.
+public func latest<T, E>(producer: SignalProducer<SignalProducer<T, E>, E>) -> SignalProducer<T, E> {
+	return SignalProducer<T, E> { sink, disposable in
+		producer.startWithSignal { outerSignal, outerDisposable in
+			disposable.addDisposable(outerDisposable)
+
+			let latestInnerDisposable = SerialDisposable()
+			disposable.addDisposable(latestInnerDisposable)
+			
+			let state = Atomic(LatestState<T, E>.initial)
+			let updateState = { (action: LatestState<T, E> -> LatestState<T, E>) -> () in
+				state.modify(action)
+				if state.value.isComplete {
+					sendCompleted(sink)
+				}
+			}
+			
+			outerSignal.observe(
+				next: { innerProducer in
+					innerProducer.startWithSignal { innerSignal, innerDisposable in
+						latestInnerDisposable.innerDisposable = innerDisposable
+						state.value = state.value.addInnerSignal(innerSignal)
+						
+						innerSignal.observe(SinkOf { event in
+							switch event {
+							case .Completed:
+								updateState { $0.completeInnerSignal(innerSignal) }
+							default:
+								state.withValue { value -> () in
+									if value.isIncompleteLatestInnerSignal(innerSignal) {
+										sink.put(event)
+									}
+								}
+							}
+						})
+					}
+				}, error: { error in
+					sendError(sink, error)
+				}, completed: {
+					updateState { $0.completeOuterSignal() }
+				})
+		}
+	}
+}
+
+private struct LatestState<T, E: ErrorType> {
+	let outerSignalComplete: Bool
+	let latestInnerSignal: LatestStateInnerSignal<T, E>
+
+	static var initial: LatestState {
+		return LatestState(
+			outerSignalComplete: false,
+			latestInnerSignal: .Complete)
+	}
+	
+	func completeOuterSignal() -> LatestState<T, E> {
+		return LatestState(
+			outerSignalComplete: true,
+			latestInnerSignal: latestInnerSignal)
+	}
+	
+	func addInnerSignal(signal: Signal<T, E>) -> LatestState<T, E> {
+		if isComplete {
+			return self
+		} else {
+			return LatestState(
+				outerSignalComplete: outerSignalComplete,
+				latestInnerSignal: .Incomplete(signal))
+		}
+	}
+	
+	func completeInnerSignal(signal: Signal<T, E>) -> LatestState<T, E> {
+		if isIncompleteLatestInnerSignal(signal) {
+			return LatestState(
+				outerSignalComplete: outerSignalComplete,
+				latestInnerSignal: .Complete)
+		} else {
+			return self
+		}
+	}
+	
+	func isIncompleteLatestInnerSignal(signal: Signal<T, E>) -> Bool {
+		switch latestInnerSignal {
+		case .Incomplete(let latestSignal) where signal === latestSignal:
+			return true
+		case .Incomplete, .Complete:
+			return false
+		}
+	}
+	
+	var isComplete: Bool {
+		switch latestInnerSignal {
+		case .Complete:
+			return outerSignalComplete
+		case .Incomplete:
+			return false
+		}
+	}
+}
+
+private enum LatestStateInnerSignal<T, E: ErrorType> {
```

Comment:
In this context I would use `latestCompleted` bool instead. You're not gaining anything from the enum and it would simplify the above cases where your forced into pattern matching.